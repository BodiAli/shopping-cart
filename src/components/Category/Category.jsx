import { Suspense } from "react";
import { Await, defer, useLoaderData, useOutletContext, useNavigate } from "react-router-dom";
import getData from "../../utils/fetchData";
import LoadingSpinner from "../LoadingSpinner/LoadingSpinner";
import generatePrice from "../../utils/generatePrice";
import styles from "./Category.module.css";

// Image
import noImage from "../../assets/no-image.svg";

// Icons
import { IoMdAdd, IoMdRemove } from "react-icons/io";

function Category() {
  const { data: games } = useLoaderData();
  const navigate = useNavigate();
  const { arr, handleButtonClick } = useOutletContext();
  const [cart] = arr;

  function navigateToProduct(gameId) {
    navigate(`/product/${gameId}`);
  }

  return (
    <Suspense fallback={<LoadingSpinner />}>
      <Await resolve={games}>
        {(games) => {
          if (!games || !games.games || games.games.length === 0) {
            throw new Error("Data not found, try again later");
          }

          return (
            <section className={styles.section}>
              {games.games.map((game) => {
                const price = generatePrice();
                game.price = price;
                game.originalPrice = price;
                game.quantity = 1;
                return (
                  <div
                    onClick={() => navigateToProduct(game.game_id)}
                    key={game.title}
                    className={styles.card}
                  >
                    <div className={styles.imageContainer}>
                      <div
                        className={styles.backGround}
                        style={{
                          backgroundImage: `url(${
                            game.sample_cover ? game.sample_cover.thumbnail_image : noImage
                          })`,
                        }}
                      ></div>
                      <img
                        src={game.sample_cover ? game.sample_cover.thumbnail_image : noImage}
                        alt={`${game.title} game poster`}
                      />
                    </div>
                    <div className={styles.textContainer}>
                      <h3>{game.title}</h3>
                      <div>
                        <p>{`Price: ${game.originalPrice}$`}</p>
                        <button onClick={(e) => handleButtonClick(e, game)}>
                          {cart.isTheProductPresent(game) ? (
                            <>
                              <IoMdRemove color="#c00000" /> Remove from Cart
                            </>
                          ) : (
                            <>
                              <IoMdAdd color="#01f5b8" /> Add to Cart
                            </>
                          )}
                        </button>
                      </div>
                    </div>
                  </div>
                );
              })}
            </section>
          );
        }}
      </Await>
    </Suspense>
  );
}

async function loader({ params }) {
  if (params.category === "games") {
    return defer({ data: getData("?") });
  } else if (params.category === "windows") {
    return defer({ data: getData("?platform=3&") });
  } else if (params.category === "mac") {
    return defer({ data: getData("?platform=74&") });
  } else if (params.category === "linux") {
    return defer({ data: getData("?platform=1&") });
  } else if (params.category === "playstation") {
    return defer({ data: getData("?platform=6&") });
  } else if (params.category === "android") {
    return defer({ data: getData("?platform=91&") });
  } else if (params.category === "iphone") {
    return defer({ data: getData("?platform=86&") });
  } else if (params.category === "action") {
    return defer({ data: getData("?genre=1&") });
  } else if (params.category === "adventure") {
    return defer({ data: getData("?genre=2&") });
  } else if (params.category === "education") {
    return defer({ data: getData("?genre=12&") });
  } else if (params.category === "puzzle") {
    return defer({ data: getData("?genre=118&") });
  } else if (params.category === "arcade") {
    return defer({ data: getData("?genre=9&") });
  } else if (params.category === "racing") {
    return defer({ data: getData("?genre=6&") });
  } else if (params.category === "simulation") {
    return defer({ data: getData("?genre=3&") });
  } else if (params.category === "strategy") {
    return defer({ data: getData("?genre=4&") });
  }
  throw new Error("Data not found");
}

Category.loader = loader;

export default Category;
