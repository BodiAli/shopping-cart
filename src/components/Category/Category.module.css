.section {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  gap: 20px;
  padding: 1rem;
  max-height: calc(100vh - 85px);
  overflow-y: scroll;
  scrollbar-color: var(--primary) var(--gray);
  position: relative;
  z-index: 0;
}

.card {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 10px;
  background-color: var(--gray);
  color: black;
  box-shadow: 2.8px 2.8px 10px rgba(0, 0, 0, 0.775);
  transition: transform 300ms ease, box-shadow 1000ms ease;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
}

.card:hover {
  transform: scale(1.01);
  box-shadow: 0px 0px 5px 10px var(--gray);
}

.imageContainer {
  position: relative;
  width: 100%;
}

.imageContainer > .backGround {
  position: absolute;
  width: 100%;
  height: 100%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  z-index: 1;
  opacity: 0.5;
  filter: blur(2px);
}

.imageContainer > img {
  width: 100%;
  height: 410px;
  position: relative;
  z-index: 2;
  object-fit: contain;
}

.textContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 1rem 0;
  gap: 10px;
}

.textContainer > h3 {
  font-size: 1.3rem;
  font-weight: 500;
  text-shadow: 0 0 2px rgba(0, 0, 0, 0.775);
  min-height: 45px;
}

.textContainer > div {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 10px;
  padding: 0.2rem 1rem;
  box-shadow: 1px 1px 5px rgba(0, 0, 0, 0.775);
  border-radius: 5px;
}

.textContainer button {
  background: var(--darkerGray);
  border: none;
  border-radius: 5px;
  padding: 0.4rem 0.8rem;
  font-weight: 600;
  cursor: pointer;
  display: flex;
  align-items: center;
  gap: 3px;
}

.textContainer button:hover {
  filter: brightness(1.2);
}

.textContainer button:active {
  filter: brightness(0.9);
}

@media (max-width: 710px) {
  .section {
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    padding: 1.2rem;
    max-height: initial;
  }

  .textContainer > h3 {
    font-size: 1.6rem;
    font-weight: 500;
    text-shadow: 0 0 2px rgba(0, 0, 0, 0.775);
    min-height: 45px;
  }
}
