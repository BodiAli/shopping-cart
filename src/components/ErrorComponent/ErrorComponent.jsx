import { useRouteError } from "react-router-dom";
import styles from "./ErrorComponent.module.css";

function ErrorComponent() {
  const error = useRouteError();

  function handleClick() {
    window.location.reload();
  }

  return (
    <section className={styles.section}>
      <h2>Unexpected error occurred!</h2>
      {error.data && <p>{error.data}</p>}
      <p>{error.statusText || error.message}</p>
      <button onClick={handleClick}>Retry</button>
    </section>
  );
}
export default ErrorComponent;
