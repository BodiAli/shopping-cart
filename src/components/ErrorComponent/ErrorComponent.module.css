.section {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;
  padding: 5rem;
  background-color: var(--green);
  min-height: 90vh;
}
.section > h2 {
  font-size: 3rem;
}
.section > p {
  font-size: 2rem;
}
.section > button {
  background-color: var(--darkerGray);
  border: none;
  border-radius: 5px;
  padding: 1rem 2rem;
  font-size: 1rem;
  cursor: pointer;
}

.section > button:hover {
  filter: brightness(1.1);
}

.section > button:active {
  filter: brightness(0.9);
}
