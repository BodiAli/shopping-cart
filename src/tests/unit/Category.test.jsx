import { render, screen } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import { vi } from "vitest";
import Category from "../../components/Category/Category";
import ErrorComponent from "../../components/ErrorComponent/ErrorComponent";

const MOCK_GAMES = {
  games: [
    {
      title: "Fifa game",
      sample_cover: {},
    },
    {
      title: "COD game",
      sample_cover: {},
    },
    {
      title: "Minecraft game",
      sample_cover: {},
    },
  ],
};
const MOCK_ACTION = {
  games: [
    {
      title: "CSGO action",
      sample_cover: {},
    },
    {
      title: "COD action",
      sample_cover: {},
    },
    {
      title: "TF2 action",
      sample_cover: {},
    },
  ],
};

Category.loader = vi.fn(({ params }) => {
  switch (params.category) {
    case "games":
      return { data: MOCK_GAMES };

    case "action":
      return { data: MOCK_ACTION };
    default:
      throw new Error("Data not found");
  }
});

const routes = [
  {
    path: "/:category",
    element: <Category />,
    errorElement: <ErrorComponent />,
    loader: Category.loader,
  },
];

function createTestWrapper(initialIndex) {
  const router = createMemoryRouter(routes, {
    initialEntries: ["/games", "/action", "/shirts"],
    initialIndex,
  });

  return function wrapper() {
    return <RouterProvider router={router} />;
  };
}
const wrapper = createTestWrapper(0);
const wrapper2 = createTestWrapper(1);
const wrapper3 = createTestWrapper(2);

describe("Category component", () => {
  test("should render correct elements", () => {
    const { container } = render(<Category />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  test("should render game titles for 'games' category", () => {
    render(<Category />, { wrapper });
    const gamesTitle = screen.getAllByRole("heading", { name: /game$/i });

    gamesTitle.forEach((title) => {
      expect(title).toBeInTheDocument();
    });
  });

  test("should render game images for 'games' category", () => {
    render(<Category />, { wrapper });
    const gameImages = screen.getAllByRole("img", { name: /game game poster$/i });

    gameImages.forEach((img) => {
      expect(img).toBeInTheDocument();
    });
  });

  test("should filter fetch games array based on params", () => {
    render(<Category />, { wrapper: wrapper2 });
    const gamesTitle = screen.getAllByRole("heading", { name: /action$/i });

    gamesTitle.forEach((title) => {
      expect(title).toBeInTheDocument();
    });
  });

  test("should display an error message when category is not found", () => {
    render(<Category />, { wrapper: wrapper3 });
    expect(screen.getByText(/data not found/i)).toBeInTheDocument();
  });
});
