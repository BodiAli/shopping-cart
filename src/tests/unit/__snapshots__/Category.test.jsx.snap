// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Category component > should render correct elements 1`] = `
<div>
  <section
    class="_section_0a39a7"
  >
    <div
      class="_card_0a39a7"
    >
      <div
        class="_imageContainer_0a39a7"
      >
        <div
          class="_backGround_0a39a7"
          style="background-image: url(undefined);"
        />
        <img
          alt="Fifa game game poster"
        />
      </div>
      <div
        class="_textContainer_0a39a7"
      >
        <h3>
          Fifa game
        </h3>
        <div>
          <p>
            Price: 50.00$
          </p>
          <button>
            Add to cart
          </button>
        </div>
      </div>
    </div>
    <div
      class="_card_0a39a7"
    >
      <div
        class="_imageContainer_0a39a7"
      >
        <div
          class="_backGround_0a39a7"
          style="background-image: url(undefined);"
        />
        <img
          alt="COD game game poster"
        />
      </div>
      <div
        class="_textContainer_0a39a7"
      >
        <h3>
          COD game
        </h3>
        <div>
          <p>
            Price: 50.00$
          </p>
          <button>
            Add to cart
          </button>
        </div>
      </div>
    </div>
    <div
      class="_card_0a39a7"
    >
      <div
        class="_imageContainer_0a39a7"
      >
        <div
          class="_backGround_0a39a7"
          style="background-image: url(undefined);"
        />
        <img
          alt="Minecraft game game poster"
        />
      </div>
      <div
        class="_textContainer_0a39a7"
      >
        <h3>
          Minecraft game
        </h3>
        <div>
          <p>
            Price: 50.00$
          </p>
          <button>
            Add to cart
          </button>
        </div>
      </div>
    </div>
    ;
  </section>
</div>
`;

exports[`Category component > should render correct elements 2`] = `<div />`;
